Source/Core/InputCommon/ControllerInterface/SteamDeck/SteamDeck.cpp

```diff
@@ -0,0 +1,302 @@
+  AddInput(new AnalogInput("Right Stick Y-",          m_latest_input.r_stick_y,    -32767));
+  AddInput(new AnalogInput("Right Stick Y+",          m_latest_input.r_stick_y,    32767));
+  AddInput(new AnalogInput("Left Pad X-",             m_latest_input.l_pad_x,      -32767));
+  AddInput(new AnalogInput("Left Pad X+",             m_latest_input.l_pad_x,      32767));
+  AddInput(new AnalogInput("Left Pad Y-",             m_latest_input.l_pad_y,      -32767));
+  AddInput(new AnalogInput("Left Pad Y+",             m_latest_input.l_pad_y,      32767));
+  AddInput(new AnalogInput("Right Pad X-",            m_latest_input.r_pad_x,      -32767));
+  AddInput(new AnalogInput("Right Pad X+",            m_latest_input.r_pad_x,      32767));
+  AddInput(new AnalogInput("Right Pad Y-",            m_latest_input.r_pad_y,      -32767));
+  AddInput(new AnalogInput("Right Pad Y+",            m_latest_input.r_pad_y,      32767));
+  AddInput(new AnalogInput("Left Pad Force",          m_latest_input.l_pad_force,  32767));
+  AddInput(new AnalogInput("Right Pad Force",         m_latest_input.r_pad_force,  32767));
+  AddInput(new AnalogInput("Left Stick Capacitance",  m_latest_input.l_stick_capa, 32767));
+  AddInput(new AnalogInput("Right Stick Capacitance", m_latest_input.r_stick_capa, 32767));
+
+  // 0x4000 LSBs = 1 g
+  // final output in m/s^2
+  constexpr auto accel_scale = 16384.0 / MathUtil::GRAVITY_ACCELERATION;
+  AddInput(new MotionInput("Accel Up",        m_latest_input.accel_z, -accel_scale));
+  AddInput(new MotionInput("Accel Down",      m_latest_input.accel_z, accel_scale));
+  AddInput(new MotionInput("Accel Left",      m_latest_input.accel_x, accel_scale));
+  AddInput(new MotionInput("Accel Right",     m_latest_input.accel_x, -accel_scale));
+  AddInput(new MotionInput("Accel Forward",   m_latest_input.accel_y, -accel_scale));
+  AddInput(new MotionInput("Accel Backward",  m_latest_input.accel_y, accel_scale));
+
+  // 16.384 (?) LSBs = 1 deg / s
+  // final output in rads / s
+  constexpr auto gyro_scale = 16.384 * 360.0 / MathUtil::TAU;
+  AddInput(new MotionInput("Gyro Pitch Up",   m_latest_input.gyro_pitch, gyro_scale));
+  AddInput(new MotionInput("Gyro Pitch Down", m_latest_input.gyro_pitch, -gyro_scale));
+  AddInput(new MotionInput("Gyro Roll Left",  m_latest_input.gyro_roll,  -gyro_scale));
+  AddInput(new MotionInput("Gyro Roll Right", m_latest_input.gyro_roll,  gyro_scale));
+  AddInput(new MotionInput("Gyro Yaw Left",   m_latest_input.gyro_yaw,   gyro_scale));
+  AddInput(new MotionInput("Gyro Yaw Right",  m_latest_input.gyro_yaw,   -gyro_scale));
+  // clang-format on
+}
+
+std::string Device::GetName() const
+{
+  return "Steam Deck";
+}
+
+std::string Device::GetSource() const
+{
+  return std::string(STEAMDECK_SOURCE_NAME);
+}
+
+void Device::UpdateInput()
+{
+  DeckInputReport rpt;
+  bool got_anything = false;
+  while (hid_read(m_device, (unsigned char*)&rpt, sizeof(rpt)) > 0)
+  {
+    got_anything = true;
```

Comment:
Shouldn't this break after the first assignment of this instead of continuing to loop?

Reply:
I don't know, should it? I've set the opened hidraw file to be nonblocking, and the Deck's controller seems to emit input reports at a fixed rate of 250 Hz, so this code consumes all available reports and returns the most recently available data (or the last available data if no data came in since the last time). This seems to be consistent with other input schemes I've looked at that work on an event queue rather than a "get the current state" mechanism (DualShockUDPClient, evdev, Pipes, Xlib). In my testing (which involved printing to the console, so it's not particularly accurate), there would often be at least two reports available (but sometimes also zero, hence this `got_anything` check existing in the first place).

All of the values in an input packet other than the motion inputs are "absolute," so dropping input reports should be fine for them. However, I'm not sure how sampling and re-sampling of motion data is supposed to work in Dolphin or on the Wii itself.

Reply:
The code's current behavior of consuming all and processing just the latest report sounds appropriate.

Reply:
added a brief comment